src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs

```diff
@@ -92,6 +92,31 @@ protected string BuildRedirectUri(string targetPath)
             return Request.Scheme + "://" + Request.Host + OriginalPathBase + targetPath;
         }
 
+        /// <summary>
+        /// Handles errors by redirecting to an error handler path if configured.  Otherwise sets the status code to 500.
+        /// </summary>
+        /// <param name="error"></param>
+        /// <returns></returns>
+        protected virtual Task HandleErrorAsync(ErrorContext error)
```

Comment:
> Yeah that's definitely one issue right now and related to the greater problem is a lot of the errors are in HandleAuthenticate which currently returns a Ticket and has no flow control... So I couldn't use this new error behavior there at all

Yup, a little refactoring will be needed to fix that since we have no way to control the execution flow from `HandleAuthenticateAsync`. The approach you used is probably the only/best one: calling `HandleErrorAsync` from `InvokeAsync`/`InvokeReturnPathAsync` when the returned ticket is null. That said, flowing the exact "error message" in this approach is a bit hard.

A related thing is to decide whether we should throw an exception from `HandleAuthenticateAsync` or return null to indicate that authentication failed. In ASOS, I opted for the second approach but I have no strong opinion on that: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/vNext/src/AspNet.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.cs#L54-L89. Opting for exceptions would allow catching them in `InvokeReturnPathAsync` and call `HandleErrorAsync` with the appropriate exception/message.

(related issue: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/issues/138#issuecomment-141719610)